### PR TITLE
[Security] Argon2i Password Encoder

### PR DESCRIPTION
**Pull request by zanbaldwin**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | WIP

Since the [libsodium RFC](https://wiki.php.net/rfc/libsodium) passed with flying colours, I'd like to kick start a discussion about adding Argon2i as a password encoder to the security component. The initial code proposal in this PR supports both the upcoming public API confirmed for PHP 7.2, and the [libsodium PECL extension](https://pecl.php.net/package/libsodium) for those below 7.2 (available for PHP 5.4+).

#### Concerns

- Should the test cover hash length? At the moment the result of Argon2i is 96 characters, but because the hashing parameters are included in the result (`$argon2i$v=19$m=32768,t=4,p=1$...`) this is not guaranteed.
- I've used one password encoder class because the result *should* be the same whether running natively in 7.2 or from the PECL extension, but should the logic be split out into separate private methods (like `Argon2iPasswordEncoder::encodePassword()`) or not (like in `Argon2iPasswordEncoder::isPasswordValid()`)? Since I can't really find anything concrete on Symfony choosing one way over another I'm assuming it's down to personal preference?

#### The Future

Whilst the libsodium RFC has been approved and the public API confirmed, there has been no confirmation of Argon2i becoming an official algorithm for `passhword_hash()`. If that is confirmed, then the implementation should *absolutely* use the native `password_*` functions since the `sodium_*` functions do not have an equivalent to the `password_needs_rehash()` function.

Any feedback would be greatly appreciated :smiley:

### CODE REVIEW

**Comment by iltar** [comment]
With the constant changing of recommended algorithms, would it be an idea to provide a migration layer in Symfony? Currently people will have to build this themselves, but it could be a lot easier if Symfony would provide an extension point, especially now that we can rehash with native php functions.

**Comment by zanderbaldwin** [comment]
That's a great idea since Symfony really should be promoting security best practices (though likely out of the scope of this PR so I'll have a go at starting that separately - if you've got any ideas on what direction that would take, what the config would look like, etc).

It's pretty much a given that Argon2i will be the next hashing algorithm supported by `password_hash()` (and therefore `password_needs_rehash()`) in PHP 7.2 but I haven't actually found anywhere that confirms it - either way it won't become `PASSWORD_DEFAULT` until at least 7.3 :cry:

**Comment by stof** [comment]
> is a new password encoder considered a new feature (therefore should be against master) or a security improvement/patch (therefore should be against 2.8)?

to me, it is a new feature

**Comment by stof** [comment]
> Is a MAX_PASSWORD_LENGTH of 72 appropriate? Argon2i is, in terms of execution duration, generally along the lines of BCrypt with a cost of 11.

the max password length of 72 for bcrypt is because the bcrypt algorithm does not support longer passwords at all (it truncates the input at 72 chars).
If Argon2i does not have a 72 chars restriction, we should not add it here (keeping the much longer restriction we have in all other encoders)

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security.xml [hunk 1]
```diff
@@ -14,6 +14,7 @@
         <parameter key="security.encoder.plain.class">Symfony\Component\Security\Core\Encoder\PlaintextPasswordEncoder</parameter>
         <parameter key="security.encoder.pbkdf2.class">Symfony\Component\Security\Core\Encoder\Pbkdf2PasswordEncoder</parameter>
         <parameter key="security.encoder.bcrypt.class">Symfony\Component\Security\Core\Encoder\BCryptPasswordEncoder</parameter>
+        <parameter key="security.encoder.argon2i.class">Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder</parameter>
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security.xml]: we don't use parameters for class names anymore

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php [hunk 2]
```diff
@@ -85,6 +86,28 @@ public function testEncodePasswordPbkdf2()
         $this->assertTrue($encoder->isPasswordValid($hash, 'password', $salt));
     }
 
+    public function testEncodePasswordArgon2i()
+    {
+        if (PHP_VERSION_ID < 70200 && !extension_loaded('libsodium')) {
+            $this->markTestSkipped('Argon2i algorithm not available.');
+
+            return;
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php]: useless. ``markTestSkipped`` is throwing an exception to interrupt the execution of the test

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function encodePassword($raw, $salt)
+    {
+        if ($this->isPasswordTooLong($raw)) {
+            throw new BadCredentialsException('Invalid password.');
+        }
+
+        if (PHP_VERSION_ID >= 70200) {
+            return $this->encodePasswordNative($raw);
+        } elseif (extension_loaded('libsodium')) {
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: just ``if``

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
Please rebase this on top of master and change the target branch of the PR

**Comment by zanderbaldwin** [comment]
All done, thanks. I'll update the `MAX_PASSWORD_LENGTH` once I can find some concrete evidence that Argon2i does not have the same length limitations as BCrypt.

**Comment by stof** [comment]
@zanderbaldwin if the algo does not explicitly says it truncates the input, it means it does not do it. you would have to look at the algo description.

**Comment by stof** [comment]
another solution to detect it experimentally: hash a very long password, and then verify it with a string changing a single char in the string. and keep doing it, by increasing the altered index little by little. Once you find a case where the altered string is accepted in verify, it means you passed over the accepted input length.
Of course, you don't need to increase 1 by 1 (unless you have days to spend on your machine for that, or a botnet to distribute the work). You can start with a high value and increase by big numbers, until you find an accepted mismatch, and then bisect the length being the max.

**Comment by zanderbaldwin** [comment]
Confirmed, I tested up to 8192 characters and there are no false-positives from truncating. More than enough for a `MAX_PASSWORD_LENGTH` of 4096.

**Comment by jedisct1** [comment]
You can safely use passwords up to 2^32 bits long.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function encodePassword($raw, $salt)
+    {
+        if ($this->isPasswordTooLong($raw)) {
+            throw new BadCredentialsException('Invalid password.');
+        }
+
+        if (PHP_VERSION_ID >= 70200) {
```
Comment (derrabus) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: I'd prefer checking for features over versions whenever possible. In this case:
```
if (function_exists('sodium_crypto_pwhash_str'))
```
What do you think?

Comment (iltar) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: I agree, feature checking is more reliable when you think of php on .net or hhvm.

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 5]
```diff
@@ -94,11 +95,16 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $encoder = $this->getContainer()->get('security.encoder_factory')->getEncoder($userClass);
         $bcryptWithoutEmptySalt = !$emptySalt && $encoder instanceof BCryptPasswordEncoder;
+        $argon2iWithoutEmptySalt = !$emptySalt && $encoder instanceof Argon2iPasswordEncoder;
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I suggest adding a a marker interface on encoders not needing the salt externally, to avoid having to know them one by one (I have the same need for detecting them in FOSUserBundle 2 to avoid generating useless salts).
what do you think @symfony/deciders ?

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php (hunk 4 above)
Comment (zanderbaldwin) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Agree. I'll add that in today.
I'm also trying to find a definitive answer on whether `PASSWORD_ARGON2I` will be implemented for `password_hash()` which would be a better solution than `sodium_crypto_passwd_str()`, so that feature test could change.

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php (hunk 5 above)
Comment (zanderbaldwin) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: +1 even though I'm not a decider; I wanted to do something like that but thought it would be adding too much for the scope of this PR.

Is that something I could implement here or separate PR?

Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: let's do it separately indeed. But the introduction of this encoder makes it worth adding such feature.

Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: Good idea!

**Comment by zanderbaldwin** [comment]
Hi, all requested changes have been made and I've finally got round to fixing the conflicts with the latest version of `master` :+1:

There will be conflicts with #21620 that I'll have to fix (most likely in this PR if #21620 gets merged first since it's been marked for `3.3`).

File: src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * @author Elnur Abdurrakhimov <elnur@elnur.pro>
+ */
+class Argon2iPasswordEncoderTest extends \PHPUnit_Framework_TestCase
+{
+    const PASSWORD = 'password';
+
+    public function setUp()
+    {
+        if (PHP_VERSION_ID < 70200 && !extension_loaded('libsodium')) {
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php]: Isn't it possible to run the tests if libsodium is missing, but the  `sodium_crypto_pwhash_str()`, `sodium_memzero()`, and `sodium_crypto_pwhash_str_verify()` functions do exist?

Comment (zanderbaldwin) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php]: Yes, absolutely. I did change from version-checking to feature-checking in `Argon2iPasswordEncoder` but I unfortunately either forgot to update it here, or it snuck in whilst rebasing (either way I missed it and shall update when I get home from work).

The `isPasswordValid()` method checks for the function that it uses, but neither `encodePassword()` or `isPasswordValid()` check for the existence of `sodium_memzero()` - is this something that should be done in case one of the check functions has been defined in user-land code? It's a pretty far off edge-case but should at least be aknowledged...

Comment (zanderbaldwin) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php]: One solution could be to check the result of `ReflectionFunctionAbstract::isUserDefined`, though would this be adding too much overhead?

**Comment by nicolas-grekas** [comment]
`src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php` should also be updated to account for argon2

**Comment by zanderbaldwin** [comment]
I've updated `EncoderFactory`, but I admit that I didn't quite comprehend the `EncoderFactoryTest` to modify it appropriately, if this needs updating can someone point me in the right direction please?

**Comment by sstok** [comment]
EncoderFactoryTest doesn't need to be updated, both bcrypt and pbkdf2 are not listed in this test-class also.

File: src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+namespace Symfony\Component\Security\Core\Tests\Encoder;
+
+use Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder;
+
+/**
+ * @author Elnur Abdurrakhimov <elnur@elnur.pro>
+ */
+class Argon2iPasswordEncoderTest extends \PHPUnit_Framework_TestCase
```
Comment (sstok) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php]: This needs to use `PHPUnit\Framework\TestCase`.

Comment (sstok) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php]: This needs to use `PHPUnit\Framework\TestCase`.

**Review by sstok — CHANGES_REQUESTED** [review]

**Comment by chalasr** [comment]
> Whilst the libsodium RFC has been approved and the public API confirmed, there has been no confirmation of Argon2i becoming an official algorithm for passhword_hash(). If that is confirmed, then the implementation should absolutely use the native password_* functions since the sodium_* functions do not have an equivalent to the password_needs_rehash() function.

Confirmed it is: PHP 7.2 introduces `PASSWORD_ARGON2I` as algo for `password_rehash()` (see https://wiki.php.net/rfc/argon2_password_hash).

@zanderbaldwin What's the status of this PR? Could you finish it? Or do you want someone to take it over? I would like to get this in.

**Comment by zanderbaldwin** [comment]
Sure, I'll get on this in the next few days.

Does anyone know a good way to determine if the constant `PASSWORD_ARGON2I` is user or core defined? I'd rather feature test than test `PHP_VERSION_ID >= 70200`.

**Comment by chalasr** [comment]
@zanderbaldwin Just tried out, using `defined('PASSWORD_ARGON2I')` should do the job.
See https://3v4l.org/N8RCV

**Comment by nicolas-grekas** [comment]
@zanbaldwin would you have time to finish this PR for the end of the week?

**Comment by zanbaldwin** [comment]
I've made most of the changes. Some points I want to clarify are:

- `PASSWORD_ARGON2I` could be user-defined in versions of PHP less than 7.2 so I also do version-testing. But I hate version checking instead of feature checking, so does anyone know of a better way of doing this?
- There are three different ways to test for the presence of Sodium, do we need them all?
  - `PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')` uses the built in `password_hash()` functions.
  - `function_exists('sodium_crypto_pwhash_str')` uses the Sodium functions that are native in 7.2.
  - `extension_loaded('libsodium')` uses the Sodium extension for PHP versions less than 7.2.

I'm thinking of getting rid of the `function_exists('sodium_crypto_pwhash_str')` check, but while testing on `php:7.2.0RC2-alpine` Docker image I've found that the constant `PASSWORD_ARGON2I` is not defined (but it works completely fine on 3v4l.org). Thoughts?

**Comment by zanbaldwin** [comment]
**Also**, GitHub's *Resolve conflicts* button says there is a conflict in [UserPasswordEncoderCommand.php](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php), but when I try to rebase `master` onto this branch I get conflicts in [CHANGELOG.md](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/CHANGELOG.md) and [UPGRADE-4.0.md](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md).

Have I really used Git so little in the last few months I've forgotten how to resolve conflicts properly? :weary:

**Comment by zanbaldwin** [comment]
Ah, I did `git rebase master` since that's the target branch.
Would that be the reason Travis is spitting out 48 failed tests?

**Comment by nicolas-grekas** [comment]
3.4 is the branch where we add features, so the target branch should also be updated.

**Comment by sstok** [comment]
Make a back-up first (`git branch 2.8-argon2i-encoder-bak`)
`git rebase --onto 3.4 master` and then push with force, don't forget to change the pull-request target branch.

**Comment by stof** [comment]
should also be rebased once https://github.com/symfony/symfony/pull/21620 is merged to benefit from it

**Comment by stof** [comment]
@zanbaldwin please rebase this as the SelfSaltingEncoderInterface is now merged.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        return $hash;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        $valid = false;
+        if (PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: `\PHP_VERSION_ID` (same for other php version checks)

**Review by chalasr — COMMENTED** [review]
Can you add a note in SecurityBundle and Security CHANGELOG files?

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        if (function_exists('sodium_crypto_pwhash_str')) {
+            return $this->encodePasswordSodiumFunction($raw);
+        }
+        if (extension_loaded('libsodium')) {
+            return $this->encodePasswordSodiumExtension($raw);
+        }
+
+        throw new \LogicException(
+            'Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.'
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Should be on one line

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php [hunk 10]
```diff
@@ -129,6 +153,23 @@ public function testEncodePasswordBcryptOutput()
         $this->assertNotContains(' Generated salt ', $this->passwordEncoderCommandTester->getDisplay());
     }
 
+    public function testEncodePasswordArgon2iOutput()
+    {
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
+        ) {
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php]: condition should be on one line.

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php [hunk 11]
```diff
@@ -69,6 +70,29 @@ public function testEncodePasswordBcrypt()
         $this->assertTrue($encoder->isPasswordValid($hash, 'password', null));
     }
 
+    public function testEncodePasswordArgon2i()
+    {
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
+        ) {
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php]: condition should be on one line.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    private function encodePasswordNative($raw)
+    {
+        $hash = password_hash($raw, PASSWORD_ARGON2I);
+
+        return $hash;
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Could return directly without an intermediate var.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        $valid = false;
+        if (PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
+            $valid = !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        } elseif (function_exists('sodium_crypto_pwhash_str_verify')) {
+            $valid = !$this->isPasswordTooLong($raw) && \sodium_crypto_pwhash_str_verify($encoded, $raw);
+            \sodium_memzero($raw);
+        } elseif (extension_loaded('libsodium')) {
+            $valid = !$this->isPasswordTooLong($raw) && \Sodium\crypto_pwhash_str_verify($encoded, $raw);
+            \Sodium\memzero($raw);
+        }
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: If there are no way to check if password is valid, we should throw an exception instead of just returning `false`. At least, that's what other encoder do as this makes debugging much easier.

File: src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class Argon2iPasswordEncoderTest extends TestCase
+{
+    const PASSWORD = 'password';
+
+    public function setUp()
+    {
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php]: should be on one line

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+            return $this->encodePasswordNative($raw);
+        }
+        if (function_exists('sodium_crypto_pwhash_str')) {
+            return $this->encodePasswordSodiumFunction($raw);
+        }
+        if (extension_loaded('libsodium')) {
+            return $this->encodePasswordSodiumExtension($raw);
+        }
+
+        throw new \LogicException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: I wonder if we cannot already (also) do these checks inside `SecurityExtension` when the config is parsed? This would exclude WTF moments at runtime.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private function encodePasswordNative($raw)
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: `private` methods should come after `public` ones

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        if (\PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
+            $valid = !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        } elseif (function_exists('sodium_crypto_pwhash_str_verify')) {
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: I would use early returns instead of `elseif`

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php (hunk 15 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: In the extension, it should be an `InvalidConfigurationException`

Comment (zanbaldwin) [reply src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Okay, I've made the changes in the Security component, I'll get onto putting extra checks in the SecurityBundle config parser :slightly_smiling_face:

**Comment by zanbaldwin** [comment]
I've made some changes, but committed them in [a separate commit](4b0d83d30a2531982e7508580b201c0c7b3c3ee5) (which can be squashed later).

- Update all exception messages regarding the Argon2i algorithm not being supported to be the same string.
- Add a [static helper method on the `Argon2iPasswordEncoder` class](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php#L23-L28), all checks throughout the Security component and SecurityBundle use the same method. If this changes, it only needs to be updated in one place.
- Added a check in the [`SecurityExtension`](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L613-L615). Is this the correct place to put it, or should the check be directly in the [`MainConfiguration`](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php#L442) class using config validation callbacks?

Feedback welcomed, as always :+1:

**Comment by xabbuh** [comment]
tests are currently failing (see the CI builds) :)

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 18]
```diff
@@ -14,6 +14,7 @@ CHANGELOG
  * deprecated HTTP digest authentication
  * deprecated command `acl:set` along with `SetAclCommand` class
  * deprecated command `init:acl` along with `InitAclCommand` class
+ * Added support for the new Argon2i password encoder.
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: minor: dot at the end should be removed

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 19]
```diff
@@ -607,6 +608,18 @@ private function createEncoder($config, ContainerBuilder $container)
             );
         }
 
+        // Argon2i encoder
+        if ('argon2i' === $config['algorithm']) {
+            if (!Argon2iPasswordEncoder::isSupported()) {
+                throw new InvalidConfigurationException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: install the libsodium extension?

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 20]
```diff
@@ -13,6 +13,7 @@ CHANGELOG
    the user will always be logged out when the user has changed between
    requests.
  * deprecated HTTP digest authentication
+ * Added a new password encoder for the Argon2i hashing algorithm.
```
Comment (fabpot) [inline src/Symfony/Component/Security/CHANGELOG.md]: same, no dot

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encoder;
+
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+
+/**
+ * Argon2iPasswordEncoder uses the Argon2i hashing algorithm provided by libsodium.
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: I would remove the "provided by libsodium" part, which is an implementation detail.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+            return $valid;
+        }
+        if (extension_loaded('libsodium')) {
+            $valid = !$this->isPasswordTooLong($raw) && \Sodium\crypto_pwhash_str_verify($encoded, $raw);
+            \Sodium\memzero($raw);
+
+            return $valid;
+        }
+
+        throw new \LogicException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: same here, "install the libsodium extension" sounds better.

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php (hunk 15 above)
Comment (fabpot) [reply src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: same here, "install the libsodium extension" sounds better.

**Review by fabpot — APPROVED** [review]
with minor text comments

**Comment by zanbaldwin** [comment]
That's because Travis installs the [`symfony/security`](https://packagist.org/packages/symfony/security) package as version `dev-master 45020b8` rather than using the Security component code of this PR.

**Comment by xabbuh** [comment]
@zanbaldwin In fact, the code is right (at least on the PHP 5.5 build), but the Argon2i algorithm is not available. Tests should reflect this possibility.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+use Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder;
+
+/**
+ * @author Zan Baldwin <hello@zanbaldwin.com>
+ */
+class Argon2iPasswordEncoderTest extends TestCase
+{
+    const PASSWORD = 'password';
+
+    public function setUp()
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php]: should be protected

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php [hunk 24]
```diff
@@ -448,6 +448,10 @@ public function testEncoders()
                 'class' => 'Symfony\Component\Security\Core\Encoder\BCryptPasswordEncoder',
                 'arguments' => array(15),
             ),
+            'JMS\FooBundle\Entity\User7' => array(
+                'class' => 'Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder',
+                'arguments' => array(),
+            ),
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php]: This is what breaks tests, as it runs on PHP 5.6. I suggest writing a dedicated test in SecurityExtensionTest, skipped if argon2i not supported.

**Review by chalasr — COMMENTED** [review]
almost good :)

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        if (\PHP_VERSION_ID >= 70200 && \defined('PASSWORD_ARGON2I')) {
+            return !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        }
+        if (function_exists('sodium_crypto_pwhash_str_verify')) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Should be fully qualified too, considering https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048#issuecomment-332784692 ?

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+            $valid = !$this->isPasswordTooLong($raw) && \sodium_crypto_pwhash_str_verify($encoded, $raw);
+            \sodium_memzero($raw);
+
+            return $valid;
+        }
+        if (extension_loaded('libsodium')) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Same

File: src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php (hunk 25 above)
Comment (stof) [reply src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: indeed, it should

Comment (zanbaldwin) [reply src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php]: Sure, will fully-qualify those two functions too :)

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php (hunk 24 above)
Comment (zanbaldwin) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php]: I'm not quite sure how to implement that - will need more details or be pointed in the right direction if I'm to implement that by end of play tomorrow, sorry!

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php]: I did it for you in https://github.com/symfony/symfony/pull/21604/commits/51be4e50cacdc44f7dd65982ea281532acd581bc :)

**Review by chalasr — APPROVED** [review]
All good

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php (hunk 24 above)
Comment (zanbaldwin) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php]: Awesome, thank you very much @chalasr! Looking at that commit, it would have taken me a long time to figure that out!

**Comment by zanbaldwin** [comment]
Should I squash my five commits?

**Comment by chalasr** [comment]
@zanbaldwin We can do it while merging, as you wish. 
@stof requested changes have been made, is this ok for you?

**Comment by nicolas-grekas** [comment]
In fact yes, squashing is needed because we can't make it when several authors are in the history.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @zanbaldwin.
